File: src/linksearch/platform_filter.py
"""Canonical platform IDs matching CandidateLink.media labels."""

from __future__ import annotations

# Must match CandidateLink.media strings from search modules.
ALLOWED_PLATFORMS: frozenset[str] = frozenset(
    {"Youtube", "YoutubeShorts", "Reddit", "Tiktok", "Facebook", "Instagram"}
)


def normalize_platform_list(raw: list[str] | None) -> list[str] | None:
    """Return None = all platforms; empty after filter = none allowed."""
    if raw is None:
        return None
    seen: set[str] = set()
    out: list[str] = []
    for x in raw:
        t = (x or "").strip()
        if t in ALLOWED_PLATFORMS and t not in seen:
            seen.add(t)
            out.append(t)
    return out


def wants_platform(platforms: list[str] | None, key: str) -> bool:
    """If platforms is None, all enabled. If empty list, nothing enabled."""
    if platforms is None:
        return True
    if len(platforms) == 0:
        return False
    return key in platforms

File: frontend/src/lib/platformFilter.ts
/** Canonical IDs — must match CandidateLink.media from search modules. */
export const ALLOWED_PLATFORMS = [
  "Youtube",
  "YoutubeShorts",
  "Reddit",
  "Tiktok",
  "Facebook",
  "Instagram",
] as const;

export type PlatformId = (typeof ALLOWED_PLATFORMS)[number];

const ALLOWED = new Set<string>(ALLOWED_PLATFORMS);

/** `undefined` = all platforms (API/client omitted). Empty = none selected. */
export function normalizePlatformList(
  raw: string[] | undefined | null
): string[] | undefined {
  if (raw === undefined || raw === null) return undefined;
  const seen = new Set<string>();
  const out: string[] = [];
  for (const x of raw) {
    const t = String(x).trim();
    if (ALLOWED.has(t) && !seen.has(t)) {
      seen.add(t);
      out.push(t);
    }
  }
  return out;
}

export function wantsPlatform(
  platforms: string[] | undefined,
  key: PlatformId
): boolean {
  if (platforms === undefined) return true;
  if (platforms.length === 0) return false;
  return platforms.includes(key);
}
